In [1]:
# Constants
NUMERICS = ('int16', 'int32', 'int64', 'float16', 'float32', 'float64')

In [2]:
#from google.colab import drive
#drive.mount('/content/gdrive')

# I'm using session storage

In [3]:
import pandas as pd
df = pd.read_csv("/content/titanic-1.csv")
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

# Descripción General

In [4]:
df.head(10)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   
5            6         0       3   
6            7         0       1   
7            8         0       3   
8            9         1       3   
9           10         1       2   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   
5                                   Moran, Mr. James    male   NaN      0   
6                            McCarthy, Mr. Timothy J    male  54.0      0   
7                     Palsson, Master. Gosta Leonard    male   2.0      3   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  
5      0            330877   8.4583   NaN        Q  
6      0             17463  51.8625   E46        S  
7      1            349909  21.0750   NaN        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [5]:
df.tail()

PassengerId  Survived  Pclass                                      Name  \
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
886    male  27.0      0      0      211536  13.00   NaN        S  
887  female  19.0      0      0      112053  30.00   B42        S  
888  female   NaN      1      2  W./C. 6607  23.45   NaN        S  
889    male  26.0      0      0      111369  30.00  C148        C  
890    male  32.0      0      0      370376   7.75   NaN        Q

In [6]:
import random
ex1, ex2, ex3 = random.sample(range(len(df)), 3)

print("Dataframe total rows: ", len(df))
df_info = pd.DataFrame(data = df.dtypes)
not_missing_values_total = df.notnull().sum()
not_missing_values_percent = round(not_missing_values_total/len(df)*100,2).astype(str)+" %"

pd.concat([df_info[0].rename("Data Type"),\
           df.T[ex1].rename("Example 1"),\
           df.T[ex2].rename("Example 2"),\
           df.T[ex3].rename("Example 3"),\
           not_missing_values_total.rename("Total Not Missing"), \
           not_missing_values_percent.rename("% of not missing values")], axis=1)

Dataframe total rows:  891


Data Type                   Example 1  \
PassengerId     int64                         203   
Survived        int64                           0   
Pclass          int64                           3   
Name           object  Johanson, Mr. Jakob Alfred   
Sex            object                        male   
Age           float64                        34.0   
SibSp           int64                           0   
Parch           int64                           0   
Ticket         object                     3101264   
Fare          float64                      6.4958   
Cabin          object                         NaN   
Embarked       object                           S   

                                                     Example 2  \
PassengerId                                                 54   
Survived                                                     1   
Pclass                                                       2   
Name         Faunthorpe, Mrs. Lizzie (Elizabeth Anne Wilkin...   
Sex                                                     female   
Age                                                       29.0   
SibSp                                                        1   
Parch                                                        0   
Ticket                                                    2926   
Fare                                                      26.0   
Cabin                                                      NaN   
Embarked                                                     S   

                      Example 3  Total Not Missing % of not missing values  
PassengerId                 649                891                 100.0 %  
Survived                      0                891                 100.0 %  
Pclass                        3                891                 100.0 %  
Name         Willey, Mr. Edward                891                 100.0 %  
Sex                        male                891                 100.0 %  
Age                         NaN                714                 80.13 %  
SibSp                         0                891                 100.0 %  
Parch                         0                891                 100.0 %  
Ticket            S.O./P.P. 751                891                 100.0 %  
Fare                       7.55                891                 100.0 %  
Cabin                       NaN                204                  22.9 %  
Embarked                      S                889                 99.78 %

**Si la variable es categórica o numérica. ¿Son correctos?**

De lo que se analiza así rápidamente con tres ejemplos las columnas tienen el tipo correcto.


In [7]:
# PassengerId, Survived, Pclass -> Object
for col in ['PassengerId', 'Survived', 'Pclass']:
    df[col] = df[col].astype(str)

df.dtypes

PassengerId     object
Survived        object
Pclass          object
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

# Descripción De Variables

## Cuantitativas

In [8]:
df.select_dtypes(include = NUMERICS).describe().T 

count       mean        std   min      25%      50%   75%       max
Age    714.0  29.699118  14.526497  0.42  20.1250  28.0000  38.0   80.0000
SibSp  891.0   0.523008   1.102743  0.00   0.0000   0.0000   1.0    8.0000
Parch  891.0   0.381594   0.806057  0.00   0.0000   0.0000   0.0    6.0000
Fare   891.0  32.204208  49.693429  0.00   7.9104  14.4542  31.0  512.3292

**¿Cuál es la tarifa más cara? ¿Cuál es la edad mínima, la edad máxima y el promedio de edad?**

- Tarifa Cara = 512.3292 
- Edad Min    = 0.42
- Edad Max    = 80
- Edad Avg    = 30 


## Cualitativas

In [9]:
df.select_dtypes(include ='object').describe(include = 'O').T

count unique                      top freq
PassengerId   891    891                        1    1
Survived      891      2                        0  549
Pclass        891      3                        3  491
Name          891    891  Braund, Mr. Owen Harris    1
Sex           891      2                     male  577
Ticket        891    681                   347082    7
Cabin         204    147                  B96 B98    4
Embarked      889      3                        S  644

**¿De qué lugar embarcaron más personas? ¿Cuántas? ¿Cuántos hombres iban en el barco?**

- Top Embarque S, 644
- Hombres 577

In [10]:
for col in ["Sex", "Survived", "Pclass", "Embarked"]:
    print(df[col].value_counts(), "\n\n")

male      577
female    314
Name: Sex, dtype: int64 


0    549
1    342
Name: Survived, dtype: int64 


3    491
1    216
2    184
Name: Pclass, dtype: int64 


S    644
C    168
Q     77
Name: Embarked, dtype: int64 




**Describe lo que observas a partir de los resultados**

En el barco iban más hombres,  sobrevivieron 342 personas, había muy pocas personas de primera clase y la mayoría eran de 3ra clase.  La mayoría hizo embarque en Southampton y la minoría en Queenstown. 

## Nulos

In [11]:
df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

# Análisis De Datos

In [12]:
# Pasajero 148
df[df['PassengerId']=='148']

PassengerId Survived Pclass                              Name     Sex  \
147         148        0      3  Ford, Miss. Robina Maggie "Ruby"  female   

     Age  SibSp  Parch      Ticket    Fare Cabin Embarked  
147  9.0      2      2  W./C. 6608  34.375   NaN        S

In [13]:
# Total pagado
df["Fare"].sum()

28693.9493

In [14]:
# Sí Sobrevivió
df[df['Survived'] == '1'].head()

PassengerId Survived Pclass  \
1           2        1      1   
2           3        1      3   
3           4        1      1   
8           9        1      3   
9          10        1      2   

                                                Name     Sex   Age  SibSp  \
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0      0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0      1   

   Parch            Ticket     Fare Cabin Embarked  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
8      2            347742  11.1333   NaN        S  
9      0            237736  30.0708   NaN        C

In [15]:
# No Sobrevivió
df[df['Survived'] == '0'].head()

PassengerId Survived Pclass                            Name   Sex   Age  \
0           1        0      3         Braund, Mr. Owen Harris  male  22.0   
4           5        0      3        Allen, Mr. William Henry  male  35.0   
5           6        0      3                Moran, Mr. James  male   NaN   
6           7        0      1         McCarthy, Mr. Timothy J  male  54.0   
7           8        0      3  Palsson, Master. Gosta Leonard  male   2.0   

   SibSp  Parch     Ticket     Fare Cabin Embarked  
0      1      0  A/5 21171   7.2500   NaN        S  
4      0      0     373450   8.0500   NaN        S  
5      0      0     330877   8.4583   NaN        Q  
6      0      0      17463  51.8625   E46        S  
7      3      1     349909  21.0750   NaN        S

In [16]:
# Porcentaje (Respecto Total Pasajeros) de muertos y superviviente
round((df["Survived"].replace(["0","1"],["% de Muertos","% de Vivos"]).value_counts() * 100 )/len(df),2).astype(str) + " %"

% de Muertos    61.62 %
% de Vivos      38.38 %
Name: Survived, dtype: object

In [17]:
# Nombres en Orden Alfabético de la Primera Clase
df[df['Pclass'] == '1']["Name"].sort_values()

730                      Allen, Miss. Elisabeth Walton
305                     Allison, Master. Hudson Trevor
297                       Allison, Miss. Helen Loraine
498    Allison, Mrs. Hudson J C (Bessie Waldo Daniels)
460                                Anderson, Mr. Harry
                            ...                       
155                        Williams, Mr. Charles Duane
351             Williams-Lambert, Mr. Fletcher Fellows
55                                   Woolner, Mr. Hugh
555                                 Wright, Mr. George
325                           Young, Miss. Marie Grice
Name: Name, Length: 216, dtype: object

In [18]:
# Edad promedio de las mujeres que viajaban en cada clase
round(df[df['Sex'] == 'female'].groupby("Pclass")["Age"].mean())

Pclass
1    35.0
2    29.0
3    22.0
Name: Age, dtype: float64

## Modificación De Datos

In [19]:
df["MenorEdad"] = (df['Age'] < 18)

# Porcentaje (Respecto Total Pasajeros) de menores y mayores de edad que sobrevivieron en cada clase
round((df.groupby(["Pclass", "MenorEdad"])["Survived"].value_counts() * 100 )/len(df),2).astype(str) + " %"

Pclass  MenorEdad  Survived
1       False      1           14.03 %
                   0            8.87 %
        True       1            1.23 %
                   0            0.11 %
2       False      0           10.66 %
                   1            7.41 %
        True       1            2.36 %
                   0            0.22 %
3       False      0           36.25 %
                   1            10.1 %
        True       0             5.5 %
                   1            3.25 %
Name: Survived, dtype: object